src/util/nondet_ifthenelse.cpp

```diff
@@ -0,0 +1,134 @@
+/*******************************************************************\
+
+Module: Nondeterministic if-then-else
+
+Author: Chris Smowton, chris@smowton.net
+
+\*******************************************************************/
+
+#include "nondet_ifthenelse.h"
+
+#include <sstream>
+
+#include <util/nondet_bool.h>
+#include <util/std_code.h>
+#include <util/std_types.h>
+#include <util/std_expr.h>
+#include <util/symbol_table.h>
+#include <util/source_location.h>
+
+// This will be unified with other similar fresh-symbol routines shortly
+static symbolt &new_tmp_symbol(
+  symbol_tablet &symbol_table,
+  const std::string &prefix,
+  const irep_idt &mode)
+{
+  static size_t temporary_counter=0;
+
+  auxiliary_symbolt new_symbol;
+  symbolt *symbol_ptr;
+
+  do
+  {
+    new_symbol.name="tmp_object_factory$"+std::to_string(++temporary_counter);
+    new_symbol.base_name=new_symbol.name;
+    new_symbol.mode=mode;
+  }
+  while(symbol_table.move(new_symbol, symbol_ptr));
+
+  return *symbol_ptr;
+}
+
+/*******************************************************************\
+
+Function: nondet_ifthenelset::begin_if
+
+  Inputs:
+
+ Outputs:
+
+ Purpose: Emits instructions and defines labels for the beginning of
+          a nondeterministic if-else diamond. Code is emitted to the
```

Comment:
It may be worth reading https://github.com/diffblue/cbmc/pull/767 front to back. Quoting the my key comment:
"It seems the code here really mixes up two concepts [...]: the code is being generated as if in the GOTO intermediate level, but it's being done using high-level language constructs. There does not seem to be any need for labels or gotos here. Just generate a code_ifthenelse with a then_case and an else_case. This avoids the static variable and synthetic labels, and will thus make the code much more concise."